platforms/allwinner-d1/d1-core/src/drivers/smhc.rs

```diff
@@ -0,0 +1,754 @@
+        } else if mint.m_cb_iw_int().bit_is_set() {
+            Some(ErrorKind::CommandBusyIllegalWrite)
+        } else if mint.m_rto_back_int().bit_is_set() {
+            Some(ErrorKind::ResponseTimeout)
+        } else if mint.m_rce_int().bit_is_set() {
+            Some(ErrorKind::ResponseCrc)
+        } else if mint.m_re_int().bit_is_set() {
+            Some(ErrorKind::Response)
+        } else if mint.m_dto_bds_int().bit_is_set() {
+            Some(ErrorKind::DataTimeout)
+        } else if mint.m_dce_int().bit_is_set() {
+            Some(ErrorKind::DataCrc)
+        } else if mint.m_dse_bc_int().bit_is_set() {
+            Some(ErrorKind::DataStart)
+        } else if mint.m_dee_int().bit_is_set() {
+            Some(ErrorKind::DataEnd)
+        } else if idst.des_unavl_int().bit_is_set() || idst.fatal_berr_int().bit_is_set() {
+            Some(ErrorKind::Dma)
+        } else {
+            None
+        }
+    }
+
+    fn advance_isr(&mut self, smhc: &smhc::RegisterBlock, num: u8) {
+        tracing::trace!(state = ?self.state, smhc = num, "SMHC{num} interrupt");
+
+        let mut needs_wake = false;
+
+        self.err = Self::error_status(smhc);
+        if let Some(err) = self.err {
+            // Clear all interrupt bits
+            smhc.smhc_rintsts.write(|w| unsafe { w.bits(0xFFFF_FFFF) });
+            smhc.smhc_idst.write(|w| unsafe { w.bits(0x3FF) });
+
+            needs_wake = true;
+        }
+
+        self.state = match self.state {
+            State::Idle => State::Idle,
+            State::WaitForCommand => {
+                if smhc.smhc_mintsts.read().m_cc_int().bit_is_set() {
+                    smhc.smhc_rintsts.write(|w| w.cc().set_bit());
+                    match self.op {
+                        SmhcOp::None => State::Idle,
+                        SmhcOp::Control => {
+                            needs_wake = true;
+                            State::Idle
+                        }
+                        SmhcOp::Read { .. } | SmhcOp::Write { .. } => State::WaitForDma,
+                    }
+                } else {
+                    self.state
+                }
+            }
+            State::WaitForDma => {
+                // TODO: better way to check for RX and TX,
+                // *normal interrupt summary* does not seem to work
+                if smhc.smhc_idst.read().bits() != 0 {
+                    smhc.smhc_idst.write(|w| {
+                        w.nor_int_sum().set_bit();
+                        w.rx_int().set_bit();
+                        w.tx_int().set_bit();
+                        w
+                    });
+                    State::WaitForDataTransfer
+                } else {
+                    self.state
+                }
+            }
+            State::WaitForDataTransfer => {
+                if smhc.smhc_mintsts.read().m_dtc_int().bit_is_set() {
+                    smhc.smhc_rintsts.write(|w| w.dtc().set_bit());
```

Comment:
should this be a `modify` instead of a `write`? the write will set the register to _only_ have the DTC bit set, is that what we want? or do we only want to touch that bit and leave the others unchanged?

if we do want to clear every other bit, a comment stating that could be nice...

Reply:
This is indeed a bit counterintuitive, but many of the interrupt registers in the D1 are "write 1 to clear".
Which means that this write sets *only* the DTC bit back to 0 and the rest are left untouched.
Using `modify` here would actually reset other bits that are also set, which is undesirable.
I'll add a general comment about it in the `advance_isr` function.

Reply:
cool, thank you! let's definitely add the comment.

in other drivers, we do generally use `modify`, as it will write back every bit that's currently 1 as a 1, which _should_ be sufficient to clear the interrupt I think --- cc @jamesmunns